File: src/nallar/ps2edit/Assets.java
//
// Source code recreated from a .class file by IntelliJ IDEA
// (powered by Fernflower decompiler)
//

package nallar.ps2edit;

import com.google.common.base.Charsets;
import com.google.common.util.concurrent.Uninterruptibles;
import lombok.val;
import nallar.ps2edit.PackFile.Entry;
import nallar.ps2edit.util.Throw;

import java.io.*;
import java.nio.file.*;
import java.util.*;
import java.util.concurrent.*;

public class Assets {
	private static final int EXPECTED_REPLACEMENT_PACK_FILE_ID = 256;
	final Map<String, Entry> nameToOrig = new HashMap<>(60);
	final Map<Integer, ArrayList<Runnable>> packToActionList = new HashMap<>(60);
	final ArrayList<PackFile> packFiles = new ArrayList<>(256);
	final Map<String, byte[]> nameToReplacement = new HashMap<>(60);
	final PackFile replacementPackFile;

	public Assets(Paths path, boolean writable) {
		File packFileDir = path.assetsDir;

		int fakePackFileNumber = 0;

		Path storedReplacementPath = path.getReplacementPackFile();
		File storedReplacement = storedReplacementPath == null ? null : storedReplacementPath.toAbsolutePath().toFile();

		File replacementPackFile;
		do {
			replacementPackFile = new File(packFileDir, String.format("Assets_%03d.pack", fakePackFileNumber++));
		} while (replacementPackFile.exists() && !replacementPackFile.equals(storedReplacement));
		--fakePackFileNumber;

		if (fakePackFileNumber > EXPECTED_REPLACEMENT_PACK_FILE_ID)
			throw new RuntimeException("Replacement pack file ID " + fakePackFileNumber + " higher than expected " + EXPECTED_REPLACEMENT_PACK_FILE_ID);
		replacementPackFile = new File(packFileDir, String.format("Assets_%03d.pack", fakePackFileNumber));

		if (writable) {
			path.setReplacementPackFile(replacementPackFile.toPath().toAbsolutePath());
			Uninterruptibles.sleepUninterruptibly(1L, TimeUnit.MILLISECONDS);

			this.replacementPackFile = new PackFile(replacementPackFile);
		} else {
			this.replacementPackFile = null;
		}


		for (int i = 0; i < fakePackFileNumber; ++i) {
			PackFile pack = new PackFile(new File(packFileDir, String.format("Assets_%03d.pack", i)));
			this.nameToOrig.putAll(pack.entryMap);
			this.packFiles.add(pack);
		}

		sanityCheck();

	}

	public static void deleteReplacement(Paths path) {
		Path replacementFilePath = path.getReplacementPackFile();
		if (replacementFilePath == null) {
			System.out.println("Replacement pack file does not exist, no need to delete it.");
			return;
		}
		File original = replacementFilePath.toFile();
		if (original.exists() && !original.delete()) {
			System.err.println("Failed to delete replacement pack file.");
			Uninterruptibles.sleepUninterruptibly(5, TimeUnit.SECONDS);
			if (original.exists() && !original.delete()) {
				original.deleteOnExit();
				throw Throw.sneaky(new IOException("Failed to delete replacement pack file"));
			}
		}
		path.setReplacementPackFile(null);
	}

	private void sanityCheck() {
		Map<String, PackFile.Entry> files = new HashMap<>();
		int i = 0;
		for (PackFile packFile : packFiles) {
			for (val entry : packFile.entryMap.entrySet()) {
				if (files.put(entry.getKey(), entry.getValue()) != null) {
					if (i == EXPECTED_REPLACEMENT_PACK_FILE_ID) {
						// Assuming this is the replacement pack file
						packFile.file.delete();
						packFile.file.deleteOnExit();
						if (replacementPackFile != null) {
							replacementPackFile.file.delete();
							replacementPackFile.file.deleteOnExit();
						}
						throw new RuntimeException("Replacement pack file not deleted properly, please try again.");
					}
					throw new RuntimeException("Duplicate entry in pack file " + i + ": " + entry.getKey());
				}
			}
			i++;
		}
	}

	public byte[] getByteData(String file) {
		byte[] rep = this.nameToReplacement.get(file);
		return rep != null ? rep : this.nameToOrig.get(file).getData();
	}

	public Map<String, PackFile.Entry> getFiles() {
		Map<String, PackFile.Entry> files = new HashMap<>();
		for (PackFile packFile : packFiles) {
			files.putAll(packFile.entryMap);
		}
		return files;
	}

	public void setByteData(String file, byte[] data) {
		this.nameToReplacement.put(file, data);
	}

	public String getStringData(String file) {
		return new String(this.getByteData(file), Charsets.ISO_8859_1);
	}

	public void setStringData(String file, String data) {
		this.setByteData(file, data.getBytes(Charsets.ISO_8859_1));
	}

	public int getNumFiles() {
		return this.nameToOrig.size();
	}

	public int getNumFilesUpdated() {
		return this.nameToReplacement.size();
	}

	public boolean addReplaceAction(final String file, final String from, final String to) {
		return this.addAction(file, () -> {
			String original = Assets.this.getStringData(file);
			if (original.contains(from)) {
				Assets.this.setStringData(file, original.replace(from, to));
			} else {
				System.err.println(file + " does not contain " + from + " to replace with " + to);
			}

		});
	}

	public boolean addAction(String file, Runnable action) {
		Entry e = this.nameToOrig.get(file);
		if (e == null) {
			System.err.println("File " + file + " not in assets to add action for.");
			return false;
		} else {
			int pack = e.getPackFile().asInt();
			ArrayList<Runnable> actions = this.packToActionList.get(pack);
			if (actions == null) {
				this.packToActionList.put(pack, actions = new ArrayList<>());
			}

			actions.add(action);
			return true;
		}
	}

	public void save() {
		if (replacementPackFile == null)
			throw new Error("Can't save Assets if created with writable=false");

		for (Map.Entry<Integer, ArrayList<Runnable>> integerArrayListEntry : this.packToActionList.entrySet()) {
			ArrayList<Runnable> replacements = integerArrayListEntry.getValue();
			PackFile e = this.packFiles.get(integerArrayListEntry.getKey());
			e.openRead();

			replacements.forEach(Runnable::run);

			e.close();
		}

		this.replacementPackFile.open();

		for (Map.Entry<String, byte[]> entry : this.nameToReplacement.entrySet()) {
			replacementPackFile.addFile(entry.getKey()).setData(entry.getValue());
		}

		this.replacementPackFile.closeForce();
	}
}

File: src/main/kotlin/net/ps2stats/edit/Assets.kt
package net.ps2stats.edit

import java.io.File
import java.io.IOException
import java.util.*
import java.util.concurrent.Executors
import java.util.function.Consumer

class Assets(path: Paths, writable: Boolean) {
    internal val nameToOrig: MutableMap<String, PackFile.Entry> = HashMap(60)
    internal val packToActionList: MutableMap<Int, MutableList<() -> Unit>> = HashMap(60)
    internal val packFiles = mutableListOf<PackFile>()
    internal val nameToReplacement: MutableMap<String, ByteArray> = HashMap(60)
    internal val replacementPackFile: PackFile?

    init {
        val packFileDir = path.assetsDir

        var fakePackFileNumber = 0

        val storedReplacement = path.replacementPackFile

        var replacementPackFile: File
        do {
            replacementPackFile = File(packFileDir, String.format("Assets_%03d.pack", fakePackFileNumber++))
        } while (replacementPackFile.exists() && replacementPackFile != storedReplacement)
        --fakePackFileNumber

        if (fakePackFileNumber > EXPECTED_REPLACEMENT_PACK_FILE_ID)
            throw RuntimeException("Replacement pack file ID $fakePackFileNumber higher than expected $EXPECTED_REPLACEMENT_PACK_FILE_ID")
        replacementPackFile = File(packFileDir, String.format("Assets_%03d.pack", fakePackFileNumber))

        if (writable) {
            path.replacementPackFile = replacementPackFile
            Thread.sleep(1)

            this.replacementPackFile = PackFile(replacementPackFile)
        } else {
            this.replacementPackFile = null
        }


        for (i in 0..fakePackFileNumber - 1) {
            val pack = PackFile(File(packFileDir, String.format("Assets_%03d.pack", i)))
            this.nameToOrig.putAll(pack.entryMap)
            this.packFiles.add(pack)
        }

        sanityCheck()

    }

    private fun sanityCheck() {
        val files = HashMap<String, PackFile.Entry>()
        for ((i, packFile) in packFiles.withIndex()) {
            for ((key, value) in packFile.entryMap) {
                if (files.put(key, value) != null) {
                    if (i == EXPECTED_REPLACEMENT_PACK_FILE_ID) {
                        // Assuming this is the replacement pack file
                        packFile.file.delete()
                        packFile.file.deleteOnExit()
                        if (replacementPackFile != null) {
                            replacementPackFile.file.delete()
                            replacementPackFile.file.deleteOnExit()
                        }
                        throw RuntimeException("Replacement pack file not deleted properly, please try again.")
                    }
                    throw RuntimeException("Duplicate entry in pack file $i: $key")
                }
            }
        }
    }

    fun getByteData(file: String): ByteArray {
        val rep = this.nameToReplacement[file]
        return rep ?: this.nameToOrig[file]?.data ?: error("Couldn't find $file")
    }

    val files: Map<String, PackFile.Entry>
        get() {
            val files = HashMap<String, PackFile.Entry>()
            for (packFile in packFiles) {
                files.putAll(packFile.entryMap)
            }
            return files
        }

    fun setByteData(file: String, data: ByteArray) {
        this.nameToReplacement.put(file, data)
    }

    fun getStringData(file: String): String {
        return String(this.getByteData(file), Charsets.ISO_8859_1)
    }

    fun setStringData(file: String, data: String) {
        this.setByteData(file, data.toByteArray(Charsets.ISO_8859_1))
    }

    val numFiles: Int
        get() = this.nameToOrig.size

    val numFilesUpdated: Int
        get() = this.nameToReplacement.size

    fun addReplaceAction(file: String, from: String, to: String) {
        addAction(file, {
            val original = this@Assets.getStringData(file)
            if (original.contains(from)) {
                this@Assets.setStringData(file, original.replace(from, to))
            } else {
                System.err.println("$file does not contain $from to replace with $to")
            }
        })
    }

    fun addAction(file: String, action: () -> Unit) {
        val e = this.nameToOrig[file] ?: error("File $file not in assets to add action for.")
        packToActionList.getOrPut(e.packFile.asInt()) { mutableListOf<() -> Unit>() }.add(action)
    }

    /**
     * Calls `callable` on each entry passed in. Works on them in the order in the PackFiles.
     * Entries in a given pack file are worked on sequentially, and pre-opened.

     * @param entries
     * *
     * @param callable
     */
    fun forEntries(entries: List<PackFile.Entry>, callable: Consumer<PackFile.Entry>) {
        val perPackEntries = mutableMapOf<PackFile, MutableList<PackFile.Entry>>()

        entries.forEach { perPackEntries.getOrPut(it.packFile) { mutableListOf() }.add(it) }

        val pool = Executors.newFixedThreadPool(1)
        for ((k, v) in perPackEntries) {
            pool.execute {
                k.openRead()
                try {
                    v.forEach { callable.accept(it) }
                } finally {
                    k.close()
                }
            }
        }
    }

    fun save() {
        if (replacementPackFile == null)
            throw Error("Can't save Assets if created with writable=false")

        for ((key, replacements) in this.packToActionList) {
            val e = this.packFiles[key]
            e.openRead()

            try {
                replacements.forEach { it() }
            } finally {
                e.close()
            }
        }

        this.replacementPackFile.open()

        try {
            for ((key, value) in this.nameToReplacement) {
                replacementPackFile.addFile(key).setData(value)
            }
        } finally {
            this.replacementPackFile.closeForce()
        }
    }

    companion object {
        private val EXPECTED_REPLACEMENT_PACK_FILE_ID = 256

        fun deleteReplacement(path: Paths) {
            val original = path.replacementPackFile
            if (original == null) {
                println("Replacement pack file does not exist, no need to delete it.")
                return
            }
            if (original.exists() && !original.delete()) {
                System.err.println("Failed to delete replacement pack file.")
                Thread.sleep(5000)
                if (original.exists() && !original.delete()) {
                    original.deleteOnExit()
                    throw IOException("Failed to delete replacement pack file")
                }
            }
            path.replacementPackFile = null
        }
    }
}
